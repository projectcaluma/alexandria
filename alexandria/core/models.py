import re
import uuid

from django.core.exceptions import ImproperlyConfigured
from django.core.validators import RegexValidator
from django.db import models
from django.utils.translation import gettext_lazy as _
from localized_fields.fields import LocalizedCharField, LocalizedTextField
from rest_framework import exceptions

from .storage_clients import client


def make_uuid():
    """Return a new random UUID value.

    This indirection is done for testing purposes, so test code can mock
    uuid.uuid4(). If we wouldn't do this, then the models would have a direct
    reference that doesn't get mocked away.

    We can't replace it with a lambda because Django Migrations can't handle them.
    """
    return uuid.uuid4()


class VisibilityMixin:
    visibility_classes = None

    @classmethod
    def visibility_queryset_filter(cls, queryset, request, **kwargs):
        if cls.visibility_classes is None:
            raise ImproperlyConfigured(
                "check that app `alexandria.core.apps.DefaultConfig` is part of your `INSTALLED_APPS`."
            )

        for visibility_class in cls.visibility_classes:
            queryset = visibility_class().filter_queryset(cls, queryset, request)

        return queryset


class PermissionMixin:
    permission_classes = None

    @classmethod
    def check_permissions(cls, request, **kwargs):
        if cls.permission_classes is None:
            raise ImproperlyConfigured(
                "check that app `alexandria.core.apps.DefaultConfig` is part of your `INSTALLED_APPS`."
            )

        for permission_class in cls.permission_classes:
            if not permission_class().has_permission(cls, request):
                raise exceptions.PermissionDenied()

    def check_object_permissions(self, request):
        for permission_class in self.permission_classes:
            if not permission_class().has_object_permission(
                self.__class__, request, self
            ):
                raise exceptions.PermissionDenied()


class BaseModel(PermissionMixin, VisibilityMixin, models.Model):
    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
    created_by_user = models.CharField(
        _("created by user"), max_length=150, blank=True, null=True
    )
    created_by_group = models.CharField(
        _("created by group"), max_length=255, blank=True, null=True
    )
    modified_at = models.DateTimeField(auto_now=True, db_index=True)
    modified_by_user = models.CharField(
        _("created by user"), max_length=150, blank=True, null=True
    )
    modified_by_group = models.CharField(
        _("created by group"), max_length=255, blank=True, null=True
    )
    metainfo = models.JSONField(_("metainfo"), default=dict)

    class Meta:
        abstract = True


class UUIDModel(BaseModel):
    """
    Models which use uuid as primary key.

    Defined as alexandria default
    """

    id = models.UUIDField(primary_key=True, default=make_uuid, editable=False)

    class Meta:
        abstract = True


class SlugModel(BaseModel):
    """
    Models which use a slug as primary key.

    Defined as alexandria default for configuration so it is possible
    to merge between developer and user configuration.
    """

    slug = models.SlugField(max_length=255, primary_key=True)

    class Meta:
        abstract = True


COLOR_RE = re.compile("^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$")
color_validator = RegexValidator(COLOR_RE, _("Enter a valid color."), "invalid")


class Category(SlugModel):
    name = LocalizedCharField(
        _("category name"), blank=False, null=False, required=False
    )
    description = LocalizedTextField(
        _("category description"), null=True, blank=True, required=False
    )
    color = models.CharField(
        max_length=18,
        default="#FFFFFF",
        validators=[color_validator],
    )
    parent = models.ForeignKey(
        "self",
        on_delete=models.CASCADE,
        blank=True,
        null=True,
        related_name="children",
    )


class TagSynonymGroup(BaseModel):
    pass


class Tag(SlugModel):
    name = models.CharField(_("tag name"), blank=False, null=False, max_length=100)
    description = LocalizedTextField(
        _("tag description"), null=True, blank=True, required=False
    )
    tag_synonym_group = models.ForeignKey(
        TagSynonymGroup,
        on_delete=models.SET_NULL,
        blank=True,
        null=True,
        related_name="tags",
    )


class Mark(SlugModel):
    name = LocalizedTextField(_("Mark name"), blank=False, null=False, max_length=100)
    description = LocalizedTextField(
        _("Mark description"), null=True, blank=True, required=False
    )


class Document(UUIDModel):
    title = LocalizedCharField(
        _("document title"), blank=True, null=True, required=False
    )
    description = LocalizedTextField(
        _("document description"), null=True, blank=True, required=False
    )
    category = models.ForeignKey(
        Category,
        on_delete=models.SET_NULL,
        null=True,
        blank=True,
        related_name="documents",
    )
    tags = models.ManyToManyField(Tag, blank=True, related_name="documents")
    marks = models.ManyToManyField(Mark, blank=True, related_name="documents")
    date = models.DateField(blank=True, null=True)

    def clone(self):
        latest_original = (
            self.files.filter(variant="original").order_by("-created_at").first()
        )
        self.pk = None
        self.save()

        # no need to clone the thumbnail, it will be auto-created
        new_original = latest_original.clone()
        new_original.document = self
        new_original.save()

        return self


class File(UUIDModel):
    ORIGINAL = "original"
    THUMBNAIL = "thumbnail"

    VARIANT_CHOICES = (
        ORIGINAL,
        THUMBNAIL,
    )
    VARIANT_CHOICES_TUPLE = (
        (variant_choice, variant_choice) for variant_choice in VARIANT_CHOICES
    )
    variant = models.CharField(
        choices=VARIANT_CHOICES_TUPLE, max_length=23, default=ORIGINAL
    )
    original = models.ForeignKey(
        "self",
        on_delete=models.CASCADE,
        null=True,
        blank=True,
        related_name="renderings",
    )
    name = models.CharField(_("file name"), max_length=255)
    document = models.ForeignKey(
        Document, on_delete=models.CASCADE, related_name="files"
    )
    checksum = models.CharField(_("checksum"), max_length=255, null=True, blank=True)

    UNDEFINED = "undefined"
    COMPLETED = "completed"
    ERROR = "error"

    STATUS_CHOICES = (
        UNDEFINED,
        COMPLETED,
        ERROR,
    )
    UPLOAD_STATUS_TUPLE = tuple(
        [(status_choice, status_choice) for status_choice in STATUS_CHOICES]
    )
    upload_status = models.CharField(
        choices=UPLOAD_STATUS_TUPLE, max_length=32, default=UNDEFINED
    )

    class Meta:
        ordering = ["-created_at"]

    def delete(self, *args, **kwargs):
        super().delete(*args, **kwargs)
        client.remove_object(self.object_name)

    @property
    def object_name(self):
        return f"{self.pk}_{self.name}"

    @property
    def upload_url(self):
        return client.upload_url(self.object_name)

    @property
    def download_url(self):
        return client.download_url(self.object_name)

    def clone(self):
        source_name = self.object_name
        self.pk = None
        self.save()
        client.copy_object(source_name, self.object_name)
        return self
