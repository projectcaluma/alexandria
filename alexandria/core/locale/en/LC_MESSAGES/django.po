# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-02-21 09:05+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: alexandria/core/api.py
msgid "(copy)"
msgstr ""

#: alexandria/core/models.py
msgid "Created by user"
msgstr ""

#: alexandria/core/models.py
msgid "Created by group"
msgstr ""

#: alexandria/core/models.py
msgid "Modified by user"
msgstr ""

#: alexandria/core/models.py
msgid "Modified by group"
msgstr ""

#: alexandria/core/models.py
msgid "Metainfo"
msgstr ""

#: alexandria/core/models.py
msgid "Please enter a valid HEX color code."
msgstr ""

#: alexandria/core/models.py
msgid "Name"
msgstr ""

#: alexandria/core/models.py
msgid "Description"
msgstr ""

#: alexandria/core/models.py
msgid "Allowed mime types"
msgstr ""

#: alexandria/core/models.py
msgid "Title"
msgstr ""

#: alexandria/core/models.py
msgid "Checksum"
msgstr ""

#: alexandria/core/validations.py
msgid "File is infected with malware."
msgstr ""

#: alexandria/core/validations.py
msgid "Malware scan had an error: "
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid "File type %(mime_type)s is not allowed in category %(category)s."
msgstr ""

#: alexandria/core/validations.py
msgid "Missing Content-Type header"
msgstr ""

#: alexandria/core/validations.py
msgid "Unknown file extension"
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match file extension %(extension)s."
msgstr ""

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match detected file content "
"%(file_content_type)s."
msgstr ""

#: alexandria/core/views.py
msgid "PDF conversion is not enabled."
msgstr ""

#: alexandria/core/views.py
msgid "\"files\" filter is mandatory!"
msgstr ""

#: alexandria/core/views.py
msgid "The \"files\" filter must consist of a comma delimited list of UUIDs!"
msgstr ""

#: alexandria/core/views.py
msgid "For downloading a file use the presigned download URL."
msgstr ""

#: alexandria/core/views.py
msgid "WebDAV is not enabled."
msgstr ""

#: alexandria/core/views.py
msgid "WebDAV is not enabled for this mime type."
msgstr ""
