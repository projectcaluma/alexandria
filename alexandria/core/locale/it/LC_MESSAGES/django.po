# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-02-21 09:05+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: alexandria/core/api.py
msgid "(copy)"
msgstr "(copia)"

#: alexandria/core/models.py
msgid "Created by user"
msgstr "Creato da utente"

#: alexandria/core/models.py
msgid "Created by group"
msgstr "Creato da gruppo"

#: alexandria/core/models.py
msgid "Modified by user"
msgstr "Modificato da utente"

#: alexandria/core/models.py
msgid "Modified by group"
msgstr "Modificato  da gruppo"

#: alexandria/core/models.py
msgid "Metainfo"
msgstr "Metadati"

#: alexandria/core/models.py
msgid "Please enter a valid HEX color code."
msgstr "Inserire un codice colore HEX valido."

#: alexandria/core/models.py
msgid "Name"
msgstr "Nome"

#: alexandria/core/models.py
msgid "Description"
msgstr "Descrizione"

#: alexandria/core/models.py
msgid "Allowed mime types"
msgstr "Tipi di file consentiti"

#: alexandria/core/models.py
msgid "Title"
msgstr "Titolo"

#: alexandria/core/models.py
msgid "Checksum"
msgstr "Checksum"

#: alexandria/core/validations.py
msgid "File is infected with malware."
msgstr "Il file è infetto da malware."

#: alexandria/core/validations.py
msgid "Malware scan had an error: "
msgstr "La scansione del malware ha avuto un errore: "

#: alexandria/core/validations.py
#, python-format
msgid "File type %(mime_type)s is not allowed in category %(category)s."
msgstr ""
"Il tipo di file %(mime_type)s non è consentito nella categoria %(category)s."

#: alexandria/core/validations.py
msgid "Missing Content-Type header"
msgstr "Intestazione Content-Type mancante"

#: alexandria/core/validations.py
msgid "Unknown file extension"
msgstr "Estensione file sconosciuta"

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match file extension %(extension)s."
msgstr ""
"Il tipo di contenuto %(content_type)s non corrisponde all'estensione "
"%(extension)s del file."

#: alexandria/core/validations.py
#, python-format
msgid ""
"Content-Type %(content_type)s does not match detected file content "
"%(file_content_type)s."
msgstr ""
"Il tipo di contenuto %(content_type)s non corrisponde al contenuto del file "
"%(file_content_type)s."

#: alexandria/core/views.py
msgid "PDF conversion is not enabled."
msgstr "La conversione PDF non è attivata."

#: alexandria/core/views.py
msgid "\"files\" filter is mandatory!"
msgstr "Il filtro \"files\" è obbligatorio!"

#: alexandria/core/views.py
msgid "The \"files\" filter must consist of a comma delimited list of UUIDs!"
msgstr ""
"Il filtro \"files\" deve essere costituito da un elenco di UUID delimitato "
"da virgole!"

#: alexandria/core/views.py
msgid "For downloading a file use the presigned download URL."
msgstr "Per scaricare un file, utilizzare l'URL di download predefinito."

#: alexandria/core/views.py
msgid "WebDAV is not enabled."
msgstr "WebDAV non è attivato."

#: alexandria/core/views.py
msgid "WebDAV is not enabled for this mime type."
msgstr "WebDAV non è abilitato per questo tipo di file."
