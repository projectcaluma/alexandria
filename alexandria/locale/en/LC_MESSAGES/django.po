# SOME DESCRIPTIVE TITLE.
# Copyright (C) YEAR THE PACKAGE'S COPYRIGHT HOLDER
# This file is distributed under the same license as the PACKAGE package.
# FIRST AUTHOR <EMAIL@ADDRESS>, YEAR.
#
#, fuzzy
msgid ""
msgstr ""
"Project-Id-Version: PACKAGE VERSION\n"
"Report-Msgid-Bugs-To: \n"
"POT-Creation-Date: 2025-02-17 16:07+0000\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\n"
"Language-Team: LANGUAGE <LL@li.org>\n"
"Language: \n"
"MIME-Version: 1.0\n"
"Content-Type: text/plain; charset=UTF-8\n"
"Content-Transfer-Encoding: 8bit\n"
"Plural-Forms: nplurals=2; plural=(n != 1);\n"

#: alexandria/core/models.py:43
msgid "Created by user"
msgstr ""

#: alexandria/core/models.py:46
msgid "Created by group"
msgstr ""

#: alexandria/core/models.py:50
msgid "Modified by user"
msgstr ""

#: alexandria/core/models.py:53
msgid "Modified by group"
msgstr ""

#: alexandria/core/models.py:55
msgid "Metainfo"
msgstr ""

#: alexandria/core/models.py:90
msgid "Please enter a valid HEX color code."
msgstr ""

#: alexandria/core/models.py:95 alexandria/core/models.py:128
#: alexandria/core/models.py:142 alexandria/core/models.py:232
msgid "Name"
msgstr ""

#: alexandria/core/models.py:97 alexandria/core/models.py:130
#: alexandria/core/models.py:144 alexandria/core/models.py:150
msgid "Description"
msgstr ""

#: alexandria/core/models.py:103
msgid "Allowed mime types"
msgstr ""

#: alexandria/core/models.py:149
msgid "Title"
msgstr ""

#: alexandria/core/models.py:236
msgid "Checksum"
msgstr ""

#: alexandria/core/validations.py:31
msgid "File is infected with malware."
msgstr ""

#: alexandria/core/validations.py:35
msgid "Malware scan had an error: "
msgstr ""

#: alexandria/core/validations.py:53
#, python-format
msgid "File type %(mime_type)s is not allowed in category %(category)s."
msgstr ""

#: alexandria/core/validations.py:71
msgid "Missing Content-Type header"
msgstr ""

#: alexandria/core/validations.py:73
msgid "Unknown file extension"
msgstr ""

#: alexandria/core/validations.py:80
#, python-format
msgid ""
"Content-Type %(content_type)s does not match file extension %(extension)s."
msgstr ""

#: alexandria/core/validations.py:92
#, python-format
msgid ""
"Content-Type %(content_type)s does not match detected file content "
"%(file_content_type)s."
msgstr ""

#: alexandria/core/views.py:120
msgid "PDF conversion is not enabled."
msgstr ""

#: alexandria/core/views.py:219
msgid "\"files\" filter is mandatory!"
msgstr ""

#: alexandria/core/views.py:231
msgid "The \"files\" filter must consist of a comma delimited list of UUIDs!"
msgstr ""

#: alexandria/core/views.py:265
msgid "For downloading a file use the presigned download URL."
msgstr ""

#: alexandria/core/views.py:280
msgid "WebDAV is not enabled."
msgstr ""

#: alexandria/core/views.py:290
msgid "WebDAV is not enabled for this mime type."
msgstr ""
