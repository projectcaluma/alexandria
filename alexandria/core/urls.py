from django.urls import path
from rest_framework.routers import SimpleRouter

from . import views

r = SimpleRouter(trailing_slash=False)

r.register(r"categories", views.CategoryViewSet)
r.register(r"documents", views.DocumentViewSet)
r.register(r"files", views.FileViewSet)
r.register(r"tags", views.TagViewSet)
r.register(r"marks", views.MarkViewSet)
r.register(r"tagsynonymgroups", views.TagSynonymGroupViewSet)

urlpatterns = r.urls

urlpatterns.append(path("hook", views.hook_view, name="hook"))
