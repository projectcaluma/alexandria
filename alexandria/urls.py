from django.conf import settings
from django.conf.urls import include
from django.conf.urls.static import static
from django.urls import path

urlpatterns = [
    path(
        "api/v1/",
        include("alexandria.core.urls"),
    ),
]

if settings.ENABLE_SILK:  # pragma: no cover
    urlpatterns.append(path("silk/", include("silk.urls", namespace="silk")))
    urlpatterns += static(settings.STATIC_URL, document_root=settings.STATIC_ROOT)
