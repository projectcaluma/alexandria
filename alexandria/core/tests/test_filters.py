import json
from itertools import combinations
from typing import Optional

import pytest
from django.urls import reverse
from rest_framework.status import HTTP_200_OK, HTTP_400_BAD_REQUEST

from ..models import Document, Tag, TagSynonymGroup
from ..views import CategoryViewSet, DocumentViewSet, FileViewSet, TagViewSet


@pytest.mark.parametrize(
    "value,status_code",
    [
        (json.dumps([{"key": "foo", "value": "bar"}]), HTTP_200_OK),
        (json.dumps([{"key": "int", "value": 5, "lookup": "gt"}]), HTTP_200_OK),
        (
            json.dumps(
                [{"key": "foo", "value": "bar"}, {"key": "baz", "value": "bla"}]
            ),
            HTTP_200_OK,
        ),
        (
            json.dumps([{"key": "foo", "value": "ar", "lookup": "contains"}]),
            HTTP_200_OK,
        ),
        (
            json.dumps([{"key": "foo", "value": "bar", "lookup": "asdfgh"}]),
            HTTP_400_BAD_REQUEST,
        ),
        (json.dumps([{"key": "foo"}]), HTTP_400_BAD_REQUEST),
        (json.dumps({"key": "foo"}), HTTP_400_BAD_REQUEST),
        ("foo", HTTP_400_BAD_REQUEST),
        ("[{foo, no json)", HTTP_400_BAD_REQUEST),
    ],
)
def test_json_value_filter(db, document_factory, admin_client, value, status_code):
    doc = document_factory(metainfo={"foo": "bar", "baz": "bla", "int": 23})
    document_factory(metainfo={"foo": "baz"})
    document_factory()
    url = reverse("document-list")
    resp = admin_client.get(url, {"filter[metainfo]": value})
    assert resp.status_code == status_code
    if status_code == HTTP_200_OK:
        result = resp.json()
        assert len(result["data"]) == 1
        assert result["data"][0]["id"] == str(doc.pk)


@pytest.mark.parametrize(
    "value,amount",
    [
        ("foo_tag", 1),
        ("_t", 2),
        ("foo_title", 1),
        ("filename.pdf", 1),
        ("filename", 2),
    ],
)
def test_document_search_filter(
    db,
    document_factory,
    tag_factory,
    file_factory,
    admin_client,
    value,
    amount,
):
    """
    Test document search filter.

    This test makes sure that search lookups are restricted to the current language.
    """
    tag1 = tag_factory(name="bar_tag")  # matches
    tag2 = tag_factory(name="foo_tag")  # doesn't match
    doc = document_factory(title={"en": "foo_title", "de": "bar_title"})  # matches
    doc2 = document_factory(
        title={"en": "bar_title", "de": "foo_title"}
    )  # doesn't match
    doc.tags.add(tag1)
    doc2.tags.add(tag2)
    file_factory(name="filename.pdf", document=doc)
    file_factory(name="filename2.pdf", document=doc2)
    document_factory()
    url = reverse("document-list")
    resp = admin_client.get(url, {"filter[search]": value})
    assert resp.status_code == HTTP_200_OK

    result = resp.json()
    assert len(result["data"]) == amount


def test_tag_category_filter(db, document_factory, tag_factory, admin_client):
    blue = tag_factory(name="blue")
    red = tag_factory(name="red")
    green = tag_factory(name="green")
    doc1 = document_factory(category__slug="cat1")
    doc1.tags.add(blue)
    doc2 = document_factory(category_id="cat1")
    doc2.tags.add(green)
    doc3 = document_factory(category__slug="cat2")
    doc3.tags.add(red)

    # one more to test distinctiveness of the result
    doc4 = document_factory(category_id="cat1")
    doc4.tags.add(blue)

    url = reverse("tag-list")
    resp = admin_client.get(url, {"filter[with-documents-in-category]": "cat1"})
    assert resp.status_code == HTTP_200_OK
    result = resp.json()
    assert len(result["data"]) == 2

    returned_tags = [tag["attributes"]["name"] for tag in result["data"]]
    assert sorted(returned_tags) == sorted(["blue", "green"])


def test_tag_document_metainfo(db, document_factory, tag_factory, admin_client):
    blue = tag_factory(name="blue")
    red = tag_factory(name="red")
    green = tag_factory(name="green")
    doc1 = document_factory(metainfo={"foo": "bar"})
    doc2 = document_factory(metainfo={"foo": "baz"})
    doc3 = document_factory(metainfo={"foo": "blah"})

    doc1.tags.add(blue)
    doc1.tags.add(red)
    doc2.tags.add(green)
    doc2.tags.add(blue)
    doc3.tags.add(green)

    url = reverse("tag-list")
    resp = admin_client.get(
        url,
        {"filter[with-documents-metainfo]": json.dumps({"key": "foo", "value": "bar"})},
    )
    assert resp.status_code == HTTP_200_OK
    result = resp.json()
    assert len(result["data"]) == 2

    returned_tags = [tag["attributes"]["name"] for tag in result["data"]]
    assert sorted(returned_tags) == ["blue", "red"]


@pytest.mark.parametrize(
    "tag_filter, expect_documents",
    [
        ("blue", ["doc1"]),
        ("green", ["doc1", "doc2"]),
        ("green,red", ["doc2"]),
        ("pink,green", ["doc1", "doc2"]),
    ],
)
def test_tag_filter(
    db, document_factory, tag_factory, tag_filter, expect_documents, admin_client
):
    blue = tag_factory(name="blue")
    red = tag_factory(name="red")
    green = tag_factory(name="green")
    pink = tag_factory(name="pink")
    doc1 = document_factory()
    doc1.tags.add(blue)
    doc1.tags.add(green)
    doc1.tags.add(pink)
    doc2 = document_factory()
    doc2.tags.add(red)
    doc2.tags.add(green)

    documents = {"doc1": doc1, "doc2": doc2}

    filter = [Tag.objects.get(name=tag).pk for tag in tag_filter.split(",")]
    url = reverse("document-list")
    resp = admin_client.get(url, {"filter[tags]": filter})
    assert resp.status_code == HTTP_200_OK
    result = resp.json()
    received_ids = set([obj["id"] for obj in result["data"]])
    expected_ids = set([str(documents[doc].pk) for doc in expect_documents])
    assert received_ids == expected_ids


def test_tag_synonym_filter(  # noqa: C901
    db, document_factory, tag_factory, admin_client, tag_synonym_group_factory
):
    url = reverse("document-list")
    ngroups = 2
    groups = {}
    groupsize = 3
    tagged_group = {}
    for i in range(1, ngroups + 1):
        group = f"synonyms{i}"
        groups[group] = None
        tagged_group[group] = [f"group{i}_syn{n}" for n in range(1, groupsize + 1)]
    ungrouped = [f"word{i}" for i in range(1, groupsize + 1)]

    def create_tagged_docs(
        synonyms: list, grouped: bool = True
    ) -> Optional[TagSynonymGroup]:
        try:
            assert len(synonyms) > 0
        except (TypeError, AssertionError) as e:  # pragma: todo cover
            raise e

        if grouped:
            tag_group = tag_synonym_group_factory(
                tags=[tag_factory(name=word) for word in synonyms]
            )
            for n in range(1, len(synonyms) + 1):
                if not tag_group.tags:  # pragma: todo cover
                    return
                perm = list(combinations(tag_group.tags.all(), n))
                for tag_set in perm:
                    document_factory(tags=tag_set)
            return tag_group
        else:
            for color in synonyms:
                document_factory(tags=[tag_factory(name=color)])

    for key, value in groups.items():
        groups[key] = create_tagged_docs(tagged_group[key])

    create_tagged_docs(ungrouped, False)

    key = list(groups.keys())[0]
    tag_sets_base = [
        doc.tags.all().values_list("id", flat=True)
        for doc in Document.objects.filter(tags__tag_synonym_group=groups[key])
    ]

    tag_group_synonyms1_tagged_documents = Document.objects.filter(
        tags__tag_synonym_group=groups[key]
    )

    for tags in tag_sets_base:
        resp = admin_client.get(url, {"filter[tags]": tags})
        assert resp.status_code == HTTP_200_OK
        result = resp.json()
        assert set(
            [str(doc.id) for doc in tag_group_synonyms1_tagged_documents]
        ) == set([obj["id"] for obj in result["data"]])

    for tag in ungrouped:
        resp = admin_client.get(url, {"filter[tags]": [Tag.objects.get(name=tag).id]})
        assert resp.status_code == HTTP_200_OK
        result = resp.json()
        assert set(
            [str(doc.id) for doc in Tag.objects.get(name=tag).documents.all()]
        ) == set([obj["id"] for obj in result["data"]])


@pytest.mark.parametrize(
    "admin_groups, active_group, expect_failure",
    [
        (["foo", "bar"], "foo", False),
        (["foo", "bar"], None, False),
        (["foo", "bar"], "somethingelse", True),
    ],
)
@pytest.mark.parametrize(
    "viewset", [CategoryViewSet, DocumentViewSet, FileViewSet, TagViewSet]
)
def test_active_group_filter(
    db, viewset, request, admin_client, active_group, expect_failure
):
    viewset_inst = viewset()
    model_name = viewset_inst.queryset.model.__name__.lower()
    records = request.getfixturevalue(f"{model_name}_factory").create_batch(2)
    url = reverse(f"{model_name}-list")

    filter = {"filter[activeGroup]": active_group} if active_group is not None else {}

    if expect_failure:
        # with pytest.raises(ValidationError):
        response = admin_client.get(url, filter)
        assert response.status_code == HTTP_400_BAD_REQUEST

    else:
        response = admin_client.get(url, filter)
        # we expect no filtering
        assert response.status_code == HTTP_200_OK
        # >= as there might be leftovers from other tests (shouldn't but may happen.)
        # Important is that the filter doesn't LIMIT its output
        assert len(response.json()["data"]) >= len(records)


@pytest.mark.parametrize(
    "mangle",
    [
        lambda name: name.lower(),
        lambda name: name.upper(),
        lambda name: name.lower()[:-3],
        lambda name: name.lower()[:-3],
        lambda name: name.upper()[:-3],
    ],
)
def test_tag_search(db, tag_factory, mangle, admin_client):
    tags = tag_factory.create_batch(4)

    search = mangle(tags[0].name)

    url = reverse("tag-list")
    resp = admin_client.get(url, {"filter[name]": search})
    assert resp.status_code == HTTP_200_OK
    result = resp.json()

    received_ids = set([obj["id"] for obj in result["data"]])
    assert str(tags[0].pk) in received_ids


@pytest.mark.parametrize(
    "has_parent,expected_count", [(True, 2), (False, 1), (None, 3)]
)
def test_has_parent(db, admin_client, category_factory, has_parent, expected_count):
    parent_category = category_factory()
    category_factory.create_batch(2, parent=parent_category)

    filters = {}
    if has_parent is not None:
        filters = {"filter[hasParent]": has_parent}

    response = admin_client.get(reverse("category-list"), filters)
    assert response.status_code == HTTP_200_OK
    result = response.json()

    received_ids = set([obj["id"] for obj in result["data"]])
    assert len(received_ids) == expected_count


@pytest.mark.parametrize(
    "filters,expected_count",
    [
        ({"filter[slug]": "category-1"}, 1),
        ({"filter[slug]": "category-none"}, 0),
        ({"filter[slugs]": "category-1,category-2"}, 2),
        ({"filter[slugs]": "category-1,category-none"}, 1),
    ],
)
def test_category_slug_filters(
    db, admin_client, category_factory, filters, expected_count
):
    category_factory(pk="category-1")
    category_factory(pk="category-2")
    category_factory(pk="category-3")

    response = admin_client.get(reverse("category-list"), filters)
    assert response.status_code == HTTP_200_OK
    assert len(response.json()["data"]) == expected_count


@pytest.mark.parametrize(
    "filters,expected_count",
    [
        ({"filter[category]": "category-1"}, 2),
        ({"filter[category]": "category-3"}, 1),
        ({"filter[category]": "category-1", "filter[exclude_children]": True}, 1),
        ({"filter[categories]": "category-none"}, 0),
        ({"filter[categories]": "category-1,category-2"}, 3),
        (
            {
                "filter[categories]": "category-1,category-2",
                "filter[exclude_children]": True,
            },
            2,
        ),
    ],
)
def test_document_category_filters(
    db,
    snapshot,
    admin_client,
    category_factory,
    document_factory,
    filters,
    expected_count,
):
    c1 = category_factory(pk="category-1")
    c2 = category_factory(pk="category-2")
    c3 = category_factory(pk="category-3", parent=c1)

    document_factory(category=c1, title="Apple")
    document_factory(category=c2, title="Pear")
    document_factory(category=c3, title="Melon")

    response = admin_client.get(reverse("document-list"), filters)

    assert response.status_code == HTTP_200_OK
    data = response.json()["data"]
    assert len(data) == expected_count
    assert sorted([doc["attributes"]["title"]["en"] for doc in data]) == snapshot
