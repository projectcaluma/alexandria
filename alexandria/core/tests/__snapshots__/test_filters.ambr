# serializer version: 1
# name: test_document_category_filters[filters0-2]
  list([
    'Apple',
    'Melon',
  ])
# ---
# name: test_document_category_filters[filters1-1]
  list([
    'Melon',
  ])
# ---
# name: test_document_category_filters[filters2-1]
  list([
    'Apple',
  ])
# ---
# name: test_document_category_filters[filters3-0]
  list([
  ])
# ---
# name: test_document_category_filters[filters4-3]
  list([
    'Apple',
    'Melon',
    'Pear',
  ])
# ---
# name: test_document_category_filters[filters5-2]
  list([
    'Apple',
    'Pear',
  ])
# ---
